python/cudf/cudf/_lib/io/utils.pyx

```diff
@@ -40,6 +44,9 @@ cdef source_info make_source_info(list src) except*:
     # Otherwise src is expected to be a numeric fd, string path, or PathLike.
     # TODO (ptaylor): Might need to update this check if accepted input types
     #                 change when UCX and/or cuStreamz support is added.
+    elif isinstance(src[0], Datasource):
+        csrc = src[0]
+        return source_info(<datasource *> csrc.c_datasource.get())
```

Comment:
I'm wondering if instead of forcing classes that inherit from `Datasource` to use a `unique_ptr[datasource]` attribute, it would be better for us to just implement a pure virtual function of something like `cdef datasource* get_datasource()` that throws for the abstract `Datasource` class which is required to be overridden for all inheriting classes.

cc @shwina if he has any opinions

Reply:
Sounds good, but does it have to be "pure virtual"? Can't it just be implemented for the base class?

```cython
cdef class Datasource:
    cdef unique_ptr[datasource] c_datasource

    cdef datasource* get_datasource(self) except *:
        return <datasource *> self.c_datasource.get()

cdef class KafkaDatasource(Datasource):
     pass
```

Perhaps I'm misunderstanding

Reply:
Will not all subclasses of `Datasource` own a `datasource` object?

Reply:
It's moreso that then forces the use of smart pointers which may be undesirable for some reason?